contracts/zns-pausable/registrar/ZNSRootRegistrarPausable.sol

```diff
@@ -0,0 +1,453 @@
+     *      - the distributionConfig HAS to be passed FULLY filled or all zeros. It is optional as a whole,
+     *      but all the parameters inside are required.
+     * @param paymentConfig (optional) Payment config for the domain to set on ZNSTreasury in the same tx
+     *  > `paymentConfig` has to be fully filled or all zeros. It is optional as a whole,
+     *  but all the parameters inside are required.
+     */
+    function registerRootDomain(
+        string calldata name,
+        address domainAddress,
+        string calldata tokenURI,
+        DistributionConfig calldata distributionConfig,
+        PaymentConfig calldata paymentConfig
+    ) external override whenNotPaused returns (bytes32) {
+        // Confirms string values are only [a-z0-9-]
+        name.validate();
+
+        // Create hash for given domain name
+        bytes32 domainHash = keccak256(bytes(name));
+
+        require(
+            !registry.exists(domainHash),
+            "ZNSRootRegistrar: Domain already exists"
+        );
+
+        // Get price for the domain
+        uint256 domainPrice = rootPricer.getPrice(0x0, name, true);
+
+        _coreRegister(
+            CoreRegisterArgs(
+                bytes32(0),
+                domainHash,
+                msg.sender,
+                domainAddress,
+                domainPrice,
+                0,
+                name,
+                tokenURI,
+                true,
+                paymentConfig
+            )
+        );
+
+        if (address(distributionConfig.pricerContract) != address(0)) {
+            // this adds additional gas to the register tx if passed
+            subRegistrar.setDistributionConfigForDomain(domainHash, distributionConfig);
+        }
+
+        return domainHash;
+    }
+
+    /**
+     * @notice External function used by `ZNSSubRegistrar` for the final stage of registering subdomains.
+     * @param args `CoreRegisterArgs`: Struct containing all the arguments required to register a domain
+     *  with ZNSRootRegistrar.coreRegister():
+     *      + `parentHash`: The hash of the parent domain (0x0 for root domains)
+     *      + `domainHash`: The hash of the domain to be registered
+     *      + `label`: The label of the domain to be registered
+     *      + `registrant`: The address of the user who is registering the domain
+     *      + `price`: The determined price for the domain to be registered based on parent rules
+     *      + `stakeFee`: The determined stake fee for the domain to be registered (only for PaymentType.STAKE!)
+     *      + `domainAddress`: The address to which the domain will be resolved to
+     *      + `tokenURI`: The tokenURI for the domain to be registered
+     *      + `isStakePayment`: A flag for whether the payment is a stake payment or not
+    */
+    function coreRegister(
+        CoreRegisterArgs memory args
+    ) external override onlyRegistrar {
```

Comment:
Should this be `whenNotPaused` as well? This `ZNSRootRegistrar` and `ZNSSubRegistrar` could hypothetically be out of sync and making this function also guarded with `whenNotPaused` would disallow subdomain registration even in the case where the subdomain registrar is not paused yet